src/storage/src/render/upsert.rs

```diff
@@ -203,6 +213,20 @@ where
                     // Find the prefix that we can emit
                     let idx = stash.partition_point(|(ts, _, _, _)| !upper.less_equal(ts));
 
+                    // Read the previous values _per key_ out of `state`, recording it
+                    // along with the value with the _latest timestamp for that key_.
+                    commands_state.clear();
+                    for (_, key, _, _) in stash.iter().take(idx) {
+                        let state_value = state.get(key).cloned();
+                        commands_state.insert(
```

Comment:
We can the entry API to avoid cloning the value multiple times here

Reply:
These are different maps, are you sure?

Reply:
Oh I see your point, one sec